firedrake/matrix.py

```diff
@@ -204,3 +209,6 @@ def __add__(self, other):
         else:
             raise TypeError("Unable to add %s to AssembledMatrix"
                             % (type(other)))
+
+    def ufl_domains(self):
+        return ufl.domain.join_domains(self.a[0].ufl_domains() + self.a[1].ufl_domains())
```

Comment:
This can be:
```
    def ufl_domains(self):
        return self._domains
````
This can be done in `MatrixBase`, and should eventually be done upstream.